### PR TITLE
Changes to PEP Calculation and Filtering

### PR DESCRIPTION
**Pull request by Alexander-Sol**
There have been several iterations of this guy, but all the tests are finally passing. See also PR 2376

 

> Currently, the pep q-value is computed by summing the pep and dividing by row number for psms or peptides sorted in descending order by MetaMorpheus score.
> 
> In this PR, pep is used to re-score the psms/peptides. PSMs are sorted in ascending order (good to bad) and then pep q-value is computed from bottom to top like we recently did for q-value using the formula (cumulative decoy count + 1) / cumulative target count).

### CODE REVIEW

**Comment by Alexander-Sol** [comment]
image (3)

These are the results from a search @trishorts ran using an arabidopsis entrapment database

**Review by trishorts — APPROVED** [review]

File: MetaMorpheus/EngineLayer/FdrAnalysis/FdrAnalysisEngine.cs [hunk 1]
```diff
@@ -47,49 +63,92 @@ private void DoFalseDiscoveryRateAnalysis(FdrAnalysisResults myAnalysisResults)
...
+                if (psms.Count > 100 & DoPEP)
                 {
-                    if (DoPEP)
+                    // Currently, inside PEP, we look at psm level Q-value when determining what should be used for training
+                    // It's not clear that this is the correct thing to do, but it's what we're doing for now
+                    CalculateQValue(psms, peptideLevelCalculation: false, pepCalculation: false);
+                    if (peptides.Count > 100 )
+                    {
+                        // I think this call is unneccesary, as peptide level q-value isn't considered in PEP
+                        CalculateQValue(peptides, peptideLevelCalculation: true, pepCalculation: false);
```
Comment (nbollis) [inline MetaMorpheus/EngineLayer/FdrAnalysis/FdrAnalysisEngine.cs:81]: Are these comments up to date?

**Review by nbollis — CHANGES_REQUESTED** [review]
Comments and questions

File: MetaMorpheus/TaskLayer/MetaMorpheusTask.cs [hunk 2]
```diff
@@ -685,14 +686,14 @@ protected void LoadModifications(string taskId, out List<Modification> variableM
...
-        protected static void WritePsmsToTsv(IEnumerable<SpectralMatch> psms, string filePath, IReadOnlyDictionary<string, int> modstoWritePruned)
+        protected static void WritePsmsToTsv(IEnumerable<SpectralMatch> psms, string filePath, IReadOnlyDictionary<string, int> modstoWritePruned, bool writePsmNotPeptideFdrInfo = true)
         {
             using (StreamWriter output = new StreamWriter(filePath))
             {
                 output.WriteLine(SpectralMatch.GetTabSeparatedHeader());
                 foreach (var psm in psms)
                 {
-                    output.WriteLine(psm.ToString(modstoWritePruned));
+                    output.WriteLine(psm.ToString(modstoWritePruned, writePsmNotPeptideFdrInfo));
```
[the remark this answers is not in the document]
Comment (Alexander-Sol) [reply MetaMorpheus/TaskLayer/MetaMorpheusTask.cs]: Yes! Good catch

**Review by trishorts — APPROVED** [review]

**Review by trishorts — APPROVED** [review]

**Review by nbollis — APPROVED** [review]

**Pull request merged**